planner/core/stats.go

```diff
@@ -150,12 +151,113 @@ func (ds *DataSource) DeriveStats(childStats []*property.StatsInfo) (*property.S
+		for _, item := range dnfItems {
+			var indexAccessPaths []*accessPath
+			if itemF, ok := item.(*expression.ScalarFunction); ok && itemF.FuncName.L == ast.LogicAnd {
+				cnfItems := expression.FlattenCNFConditions(itemF)
+				indexAccessPaths = ds.buildAccessPath(cnfItems, usedIndexCount)
+			} else {
+				tempArgs := []expression.Expression{item}
+				indexAccessPaths = ds.buildAccessPath(tempArgs, usedIndexCount)
+			}
+			if indexAccessPaths == nil {
+				ixMergePartialIxPaths = nil
+				break
+			}
+			ixMergePartialIxPath := ds.generateIndexMergePartialIndexPath(indexAccessPaths)
+			if ixMergePartialIxPath == nil {
+				ixMergePartialIxPaths = nil
+				break
+			}
+			ixMergePartialIxPaths = append(ixMergePartialIxPaths, ixMergePartialIxPath)
+		}
+		if len(ixMergePartialIxPaths) > 1 {
+			possiblePath := ds.generateIndexMergeOrPath(ixMergePartialIxPaths, i)
+			if possiblePath != nil {
+				ds.possibleAccessPaths = append(ds.possibleAccessPaths, possiblePath)
+			}
+		}
+	}
+}
+
+// buildAccessPath generates all possible index paths for conditions.
+func (ds *DataSource) buildAccessPath(conditions []expression.Expression, usedIndexCount int) []*accessPath {
+	var results = make([]*accessPath, 0, 0)
+	for i := 1; i <= usedIndexCount; i++ {
+		newAccessPath := new(accessPath)
+		newAccessPath.index = ds.possibleAccessPaths[i].index
+		_, err := ds.deriveIndexPathStats(newAccessPath, conditions)
+		if err != nil {
+			return nil
+		}
+		// if accessConds is empty or tableFilter is not empty, we ignore the access path.
+		if len(newAccessPath.tableFilters) > 0 || len(newAccessPath.accessConds) == 0 {
+			continue
+		}
+		results = append(results, newAccessPath)
+	}
+	if len(results) == 0 {
+		results = nil
+	}
+	return results
+}
+
+// generateIndexMergePartialIndexPath chooses a best index path from all possible index paths.
+// Now we just choose the index with more columns.
+func (ds *DataSource) generateIndexMergePartialIndexPath(indexAccessPaths []*accessPath) *accessPath {
+	if len(indexAccessPaths) == 1 {
+		return indexAccessPaths[0]
+	}
+
+	maxColsIndex := 0
+	maxCols := len(indexAccessPaths[0].idxCols)
+	for i := 0; i < len(indexAccessPaths); i++ {
+		current := len(indexAccessPaths[i].idxCols)
+		if current > maxCols {
+			maxColsIndex = i
+			maxCols = current
+		}
+	}
+	return indexAccessPaths[maxColsIndex]
+}
+
+// generateIndexMergeOrPath generates one possible IndexMergePath
+func (ds *DataSource) generateIndexMergeOrPath(ixMergePartialIxPaths []*accessPath, current int) *accessPath {
+	indexMergePath := new(accessPath)
+	indexMergePath.isIndexMergePath = true
+	indexMergePath.indexMergeType = 3
+	indexMergePath.partialIndexPaths = append(indexMergePath.partialIndexPaths, ixMergePartialIxPaths...)
+
+	for i := 0; i < len(ds.pushedDownConds); i++ {
+		if i == current {
+			continue
+		}
+		indexMergePath.tableFilters = append(indexMergePath.tableFilters, ds.pushedDownConds[i])
+	}
+	return indexMergePath
```

Comment:
You don't need this loop, try this:
```golang
path.tableFilters = append(ds.pushedDownConds[:ordinal], ds.pushedDownConds[ordinal+1:])
```

Reply:
this way has side effects which will change the value of ds.pushedDownConds

Reply:
Then you can use:
```golang
path.tableFilters = append(path.tableFilters, ds.pushedDownConds[:ordinal])
path.tableFilters = append(path.tableFilters, ds.pushedDownConds[ordinal+1:])
```